powerauth-push-server/src/main/java/io/getlime/push/service/PushMessageSenderService.java

```diff
@@ -332,19 +372,55 @@ private void sendMessageToAndroid(final FcmClient fcmClient, final PushMessageBo
             request.setNotification(notification);
         }
-        final ListenableFuture<ResponseEntity<String>> future = fcmClient.exchange(request);
+        final ListenableFuture<ResponseEntity<FcmSendResponse>> future = fcmClient.exchange(request);
 
-        future.addCallback(new ListenableFutureCallback<ResponseEntity<String>>() {
+        future.addCallback(new ListenableFutureCallback<ResponseEntity<FcmSendResponse>>() {
             @Override
             public void onFailure(Throwable throwable) {
                 Logger.getLogger(PushMessageSenderService.class.getName()).log(Level.SEVERE, "Notification rejected by the FCM gateway:" + throwable.getLocalizedMessage());
                 Logger.getLogger(PushMessageSenderService.class.getName()).log(Level.SEVERE, throwable.getLocalizedMessage());
-                callback.didFinishSendingMessage(PushSendingCallback.Result.FAILED);
+                callback.didFinishSendingMessage(PushSendingCallback.Result.FAILED, null);
             }
+
             @Override
-            public void onSuccess(ResponseEntity<String> stringResponseEntity) { // TODO: Implement processing of the response body
-                Logger.getLogger(PushMessageSenderService.class.getName()).log(Level.INFO, stringResponseEntity.getBody());
-                callback.didFinishSendingMessage(PushSendingCallback.Result.OK);
+            public void onSuccess(ResponseEntity<FcmSendResponse> stringResponseEntity) {
+                for (FcmResult fcmResult : stringResponseEntity.getBody().getFcmResults()) {
+                    //no issues, straight sending
+                    if (stringResponseEntity.getBody().getFailure() + stringResponseEntity.getBody().getCanonicalIds() == 0) {
+                        Logger.getLogger(PushMessageSenderService.class.getName()).log(Level.INFO, "Notification sent");
+                        callback.didFinishSendingMessage(PushSendingCallback.Result.OK, null);
+                    } else {
+                        //message sent, token has to be updated (stored in map under key "updateToken")
+                        if (fcmResult.getRegistrationId() != null) {
+                            Logger.getLogger(PushMessageSenderService.class.getName()).log(Level.INFO, "Notification sent and token has been updated");
+                            Map<String, Object> contextData = new HashMap<>();
+                            contextData.put("updateToken", fcmResult.getRegistrationId());
+                            callback.didFinishSendingMessage(PushSendingCallback.Result.OK, contextData);
+                            stringResponseEntity.getBody().setFailure(stringResponseEntity.getBody().getCanonicalIds() - 1);
+                        } else {
+                            switch (fcmResult.getFcmError()) {
+                                //token doesn't exist, remove device registration
+                                case "NotRegistered":
```

Comment:
... this is more a code style issue, but I suggest using "blocked format" of the case statements, like:

```java
case SOMETHING: {
    code.method();
    break;
}
```

This is generally required in case new variables needs to be declared (they become block scoped).